src/frameworks/opencl/api/context.rs

```diff
@@ -49,4 +54,136 @@ impl API {
             _ => Err(Error::Other("Unable to create context"))
         }
     }
+
+    /// Gets info about one of the available properties of an OpenCL context.
+    pub fn get_context_info(
+        context: cl::context_id,
+        query: ContextInfoQuery ,
+    ) -> Result<ContextInfo, Error> {
+
+        let mut info : cl::context_info = match query {
+            ContextInfoQuery::ReferenceCount => cl::CL_CONTEXT_REFERENCE_COUNT,
+            ContextInfoQuery::NumDevices => cl::CL_CONTEXT_NUM_DEVICES,
+            ContextInfoQuery::Properties => cl::CL_CONTEXT_PROPERTIES,
+            ContextInfoQuery::Devices => cl::CL_CONTEXT_DEVICES,
+        };
+
+        Ok({
+            unsafe {
+                let mut zero: usize = 0;
+                let info_size: *mut usize = &mut zero;
+                let info_ptr: *mut libc::c_void = ptr::null_mut();
+                API::ffi_get_context_info_size(context, info, info_size)
+                    .and_then(|_| {
+                        API::ffi_get_context_info(context,
+                                                  info,
+                                                  *info_size,
+                                                  info_ptr)
+                    }).and_then(|_| {
+                        match info {
+                            cl::CL_CONTEXT_REFERENCE_COUNT => {
+                                Ok(ContextInfo::ReferenceCount(info_ptr as u32))
+                            },
+                            cl::CL_CONTEXT_DEVICES => {
+                                let len = *info_size / size_of::<cl::uint>();
+                                // TODO can we really consume the output here?
+                                // TODO Vec takes over ownership
+                                let dev_ids = Vec::from_raw_parts(
+                                        info_ptr as *mut cl::uint,
+                                        len, len
+                                );
+                                Ok(ContextInfo::Devices(
+                                    dev_ids
+                                        .iter()
+                                        .map(|&id| Device::from_isize(id as isize))
+                                        .collect()
+                                ))
+                            },
+                            cl::CL_CONTEXT_NUM_DEVICES => {
+                                Ok(ContextInfo::NumDevices(info_ptr as u32))
+                            },
+                            cl::CL_CONTEXT_PROPERTIES => {
+                                let mut v : Vec<ContextProperties> = vec!();
+                                let mut start : *mut u8 = info_ptr as *mut u8;
+                                let old : *mut u8 = start.clone();
+                                loop {
```

Comment:
Depends, I think we should do that on demand and not start factoring functions out before we need them, since generalization tends to add more complexity.